### DISCUSSION TITLE
'Attach to road' performance

### DISCUSSION THREAD

**Discussion by Yaqana** [osmandapp/OsmAnd#19642] [Q&A] [answered]
I explore some new multi-day hiking trails and I often import tracks as GPX from hiking websites. They are always made of many points connected with straight lines. To use the full potential of OsmAnd I try to attach them to roads. I have never successfully attached to road any of them. It just seems to take forever. I just waited an hour for a 1.3k points track and nothing happened. I tested with smaller tracks and even 14 points took a few seconds. 100 points probably less than a minute. 
The OsmAnd website notes that it can take very long time if there are many points, but it doesn't really say how much is 'many'. Is importing a track with this many points unusual and it's normal attaching it to road never ends? Maybe the user guide (which I think is great btw) could be a bit more precise about this limitation?

[You can find an example track here](https://www.walkingenglishman.com/ldp/1066haroldsway.html#google_vignette)

I'm using iOS.

**Comment by Sonwon1** [comment]
I always look for a different import format to avoid using 'Attach to road' because I find the performance very poor.

**Reply by DmitryAlexei** [reply, marked as answer]
@Yaqana Your track is made for pedestrian profile, and it is 153 km long (this is quite a lot!). 
For long tracks with huge number of points Attach to roads option may work slowly. Please refer to our documentation https://osmand.net/docs/user/navigation/setup/gpx-navigation/#attach-to-roads
As a workaround, you can increase the threshold distance. You can also cut your track into smaller pieces and attach them one by one.

**Reply by Yaqana** [reply]
@DmitryAlexei thanks for reply! I was just curious if it's expected or it's a bug. Using "attach to road" on long tracks would be the simplest way for my case, but if the app is not meant to do that then I can live without it

**Reply by DmitryAlexei** [reply]
Instead of using paid route planner apps, we recommend using OsmAnd web https://osmand.net/map/  
Tracks made inside OsmAnd do not need to be attached to roads

**Reply by Sonwon1** [reply]
> Instead of using paid route planner apps, we recommend using OsmAnd web https://osmand.net/map/ Tracks made inside OsmAnd do not need to be attached to roads

I have several thoughts after trying to use an imported file for a road trip.

1. It is simply unusable in my opinion for road tours of 200 - 300 or more kilometers.. So I believe the documentation should reflect the current state Snap to Road is in. Something like this feature is not recommended for trips longer than 'X' kilometers or with more than 'X' waypoints. 

2. Perhaps a warning should popup when the user attempts to use the feature when it is know to be a problem and ask the user if they would like to continue.

3. I tested importing a file that failed in OsmAnd using Locus Map and the file imported without the same problems as OsmAnd?

4. Why can't OsmAnd look at all of the waypoints (not the ones generated every so many meters) in a GPS file and build the route from the waypoints? I realize the route may not be exactly the same. Couldn't that be a config question asking which import would you like to use, exact or based on waypoints?

5. Is https://osmand.net/map/ fully functional now? Can I view POI's now?

**Comment by vshcherb** [comment]
Significant performance improvement is achieved with new attaching algorithm, please give it a try with 4.8 nightly build

**Reply by 0709wiwiwi** [reply]
@vshcherb 

I just do test the attach to road performance on a track import.
I am not an osmand user but this complex feature interests me.
- A quality track record ensures a good result.
[Record_ACE2_orig.gpx.txt](https://github.com/user-attachments/files/15759630/Record_ACE2_orig.gpx.txt)
The by osmand web shown turn instructions are correct.

- The Harolds walk. 
The track points are sometimes quite (too) far apart.
OsmAnd snaps trkpt to a known osm location and thats it.
Routeyou web rerouting fills the ‘gaps’ and includes the TBT in the export gpx file.
 _(The TBT added value with this (city) walk is of little use imo)_

Conclusions so far: 
- Import junk then junk comes out.
- A manual redesign by the track shown as a template is sometimes faster.

**Reply by Sonwon1** [reply]
> @vshcherb
> 
> I just do test the attach to road performance on a track import. I am not an osmand user but this complex feature interests me.
> 
> * A quality track record ensures a good result.
>   [Record_ACE2_orig.gpx.txt](https://github.com/user-attachments/files/15759630/Record_ACE2_orig.gpx.txt)
>   The by osmand web shown turn instructions are correct.
> * The Harolds walk.
>   The track points are sometimes quite (too) far apart.
>   OsmAnd snaps trkpt to a known osm location and thats it.
>   Routeyou web rerouting fills the ‘gaps’ and includes the TBT in the export gpx file.
>   _(The TBT added value with this (city) walk is of little use imo)_
> 
> Conclusions so far:
> 
> * Import junk then junk comes out.
> * A manual redesign by the track shown as a template is sometimes faster.

What is the definition of a "quality track"? I mean what makes one track bad and another "quality", are there any specific parameters? Is that in the documentation, if not maybe it should be there?

**Reply by vshcherb** [reply]
There is a threshold parameter how close roads should be attached

**Reply by 0709wiwiwi** [reply]
>  What is the definition of a "quality track"? 

_Well...what is a good quality (recorded) track that you want to use to reproduce a navigable track ?_ 

A good track record runs steadily (timestamps). Does not include unexpected deviations. A snap to street mechanism definitely has a hard time with that.

**Reply by Sonwon1** [reply]
Thanks for confirming you were only referring to recorded tracks. My focus is on importing tracks created with map software.

**Reply by 0709wiwiwi** [reply]
You wish to  use your tracks inclusive generated TBT ? 
- What routing service do you use as your track source ? 
- Locus offers no exact map matching tool.
  Offers navigation based on the trackshape.
- What waypoints (a gpx track + wpt) are you refering too ?  Source ?

**Reply by Sonwon1** [reply]
"You wish to use your tracks inclusive generated TBT ?" Yes

"What routing service do you use as your track source ?" Furkot.com

Furkot can create OsmAnd specific GPS files with low or high resolution. I included both and I am not clear on the exact differences. I suppose high just includes more points. In the GPS file are points and waypoints. I would think OsmAnd would be able to create a route based on just waypoints (I realize the routing may not be exact).

[Forest_loop_-_3_low.gpx.txt](https://github.com/user-attachments/files/15768869/Forest_loop_-_3_low.gpx.txt)
[Forest_loop_-_3_high.gpx.txt](https://github.com/user-attachments/files/15768920/Forest_loop_-_3_high.gpx.txt)

"You expect to use this for longer distance trajects ? Motorcyle ?" Yes to both, my trips per day vary from 150 kilometers to 350 kilometers, depending on the time available and road types. 

Okay, so noted on Locus Maps.

"What waypoints (a gpx track + wpt) are you refering too ? Source ?" Source is in files about. Waypoints are points I add to a route to shape the route to go where I want. Points are added when the GPS file is created to give an exact route. I am not familiar enough with GPS files to know if there is a distinction in the file (Are waypoints and points formatted the same in the file)?

I understand exact routing is not possible unless you follow all of the waypoints and I am okay with the route not being exact. I can use the import route as is or make modifications to the route.

The problems I noted on the last trip with an imported file:

1. micro turns, OsmAnd telling me to make turns for curves in the road.
2. While on the Autobahn several times the navigation instructions want me to take an exit and then get back on the Autobahn. In several other cases at overlapping Autobahns it route me to take all the Autobahn exits until I had ridden a clover leaf. In both cases it should have kept me on the Autobahn without exiting.
3. The whole route had no road names.

That is all I remember at the moment.

**Reply by Sonwon1** [reply]
In the GPS files I posted the waypoints are labeled 'wpt'.
Track points are labeled 'trkpt'.

High resolution just includes more track points.

**Reply by Sonwon1** [reply]
"After import, a (compatible) app does not need to add extra like TBT.
Everything needed is already there, exactly as the user designed it."

When I follow the posted file with Simulated Navigation I am missing road names? Also no turn instructions at the round about? While I can see the route on the map it is a driving distraction to have to keep watching the screen for every turn. It also gives directions in a curve where there is no turn available (micro turn instructions)? This is not what I would expect. The file does not give proper turn by turn instructions in my opinion.

If I manually add the route to OsmAnd then I get road names, I get proper instructions at roundabouts and there are no micro turn instructions. This is what I would expect from an imported GPX file.

**Reply by Sonwon1** [reply]
@0709wiwiwi 

I am not trying to promote any app! I am only trying to find out why the GPX file import process is not working so that OsmAnd is more usable for me and other users. And may I politely remind you that you asked me for the app name. I quote from your post, "What routing service do you use as your track source ?" 

I would like to work with OsmAnd in resolving what is a problem for many users. I can look through previous post here and find many post related to the problem I am reporting. I would like to find a solution. Perhaps the exported GPX file from the routing service is corrupt or perhaps the import algorithm has a bug?

Maybe I am the problem? I expect GPX files to be transferable between routing and navigation services. By this I mean not just the route but also the navigation instructions should work. Is the GPX file format so loosely defined it makes the files incompatible or only partially compatible between services? I know some major navigation companies have their own GPX file format. Shame on them! If the GPX format is so broken then shouldn't this be in the documentation so users would know what the limitations are?

Here is what I am going to do. I will create a route in OsmAnd and save the route. I will also create the exact same route with a routing service and also save. I will then export both to GPX 1.1 file format (If I remember correctly the OsmAnd documentation says OsmAnd is compatible with GPX version 1.1). I will import both into OsmAnd and compare their navigation behavior. I will also compare file formats to ascertain the difference and report back here in a new thread.

Does this sound like a good plan? If you have any constructive suggestions for my plan I would like to hear them?

**Reply by 0709wiwiwi** [reply]
The definition of gpx1.1 (Y2004) does not specify how TBT instructions should be added. So a performant uniform file exchange mode between apps is unfortunately not there. Such so do result in just a lot of chaos and fumbling.  I do show  you a very simple but decent **track** file  _(yes track, no typo)_ exchange method inclusive the TBT and via points. Anyway, I don't decide what the developers of OsmAnd will or will not eventually ever implement.

**Reply by mi-hol** [reply]
> Anyway, I don't decide what the developers of OsmAnd will or will not eventually ever implement.

@Sonwon1 maybe a better approach might be to involve yourself in specific issue conversations with the label "and-gpx-track"

https://github.com/osmandapp/OsmAnd/issues?q=is%3Aopen+is%3Aissue+label%3Aand-gpx-track
